jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferAggregator.java

```diff
@@ -0,0 +1,138 @@
+ */
+public class ByteBufferAggregator
+{
+    private static final Logger LOG = LoggerFactory.getLogger(ByteBufferAggregator.class);
+
+    private final ByteBufferPool _bufferPool;
+    private final boolean _direct;
+    private final int _maxSize;
+    private RetainableByteBuffer _retainableByteBuffer;
+    private int _aggregatedSize;
+    private int _currentSize;
+
+    /**
+     * Creates a ByteBuffer aggregator.
+     * @param bufferPool The {@link ByteBufferPool} from which to acquire the buffers
+     * @param direct whether to get direct buffers
+     * @param startSize the starting size of the buffer
+     * @param maxSize the maximum size of the buffer which must be greater than {@code startSize}
+     */
+    public ByteBufferAggregator(ByteBufferPool bufferPool, boolean direct, int startSize, int maxSize)
+    {
+        if (maxSize <= 0)
+            throw new IllegalArgumentException("maxSize must be > 0, was: " + maxSize);
+        if (startSize <= 0)
+            throw new IllegalArgumentException("startSize must be > 0, was: " + startSize);
+        if (startSize > maxSize)
+            throw new IllegalArgumentException("maxSize (" + maxSize + ") must be >= startSize (" + startSize + ")");
+        _bufferPool = (bufferPool == null) ? new ByteBufferPool.NonPooling() : bufferPool;
+        _direct = direct;
+        _maxSize = maxSize;
+        _currentSize = startSize;
+    }
+
+    /**
+     * Copies the given ByteBuffer into this aggregator. This will aggregate bytes up to the
+     * specified maximum size, at which time this method returns false and
+     * {@link #takeRetainableByteBuffer()} must be called for this method to accept aggregating again.
+     * @param buffer the buffer to copy into this aggregator; its position is updated according to
+     * the number of aggregated bytes
+     * @return true if the aggregator's buffer is full and should be taken, false otherwise
+     */
+    public boolean aggregate(ByteBuffer buffer)
+    {
+        ensureBufferCapacity(buffer.remaining());
+        if (_retainableByteBuffer == null)
+        {
+            _retainableByteBuffer = _bufferPool.acquire(_currentSize, _direct);
+            BufferUtil.flipToFill(_retainableByteBuffer.getByteBuffer());
+        }
+        int prevPos = buffer.position();
+        int copySize = Math.min(_currentSize - _aggregatedSize, buffer.remaining());
+
+        _retainableByteBuffer.getByteBuffer().put(_retainableByteBuffer.getByteBuffer().position(), buffer, buffer.position(), copySize);
+        _retainableByteBuffer.getByteBuffer().position(_retainableByteBuffer.getByteBuffer().position() + copySize);
+        buffer.position(buffer.position() + copySize);
+        _aggregatedSize += buffer.position() - prevPos;
+        return _aggregatedSize == _maxSize;
+    }
+
+    private void ensureBufferCapacity(int remaining)
```

Comment:
I'd rename the method, as it does not ensure the buffer capacity.

`tryExpandBufferCapacity()` is closer to what it does.

Reply:
Ok with that name, renaming done.

Reply:
Done.